staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go

```diff
@@ -0,0 +1,212 @@
+		}
+	}
+
+	clk.Run(nil)
+	if expectedAllRequests && failedCount > 0 {
+		t.Errorf("Expected all requests to be successful but got %v failed requests", failedCount)
+	} else if !expectedAllRequests && failedCount == 0 {
+		t.Errorf("Expected failed requests but all requests succeeded")
+	}
+}
+
+// TestNoRestraint should fail because the dummy QueueSet exercises no control
+func TestNoRestraint(t *testing.T) {
+	now := time.Now()
+	clk, counter := clock.NewFakeEventClock(now, 0, nil)
+	nrf := test.NewNoRestraintFactory()
+	config := fq.QueueSetConfig{}
+	nr, err := nrf.NewQueueSet(config)
+	if err != nil {
+		t.Fatalf("QueueSet creation failed with %v", err)
+	}
+	exerciseQueueSetUniformScenario(t, nr, []uniformClient{
+		{1001001001, 5, 10, time.Second, time.Second},
+		{2002002002, 2, 10, time.Second, time.Second / 2},
+	}, time.Second*10, false, true, clk, counter)
+}
+
+func TestUniformFlows(t *testing.T) {
+	now := time.Now()
+
+	clk, counter := clock.NewFakeEventClock(now, 0, nil)
+	qsf := NewQueueSetFactory(clk, counter)
+	config := fq.QueueSetConfig{
+		Name:             "TestUniformFlows",
+		ConcurrencyLimit: 100,
+		DesiredNumQueues: 128,
+		QueueLengthLimit: 128,
+		HandSize:         1,
+		RequestWaitLimit: 10 * time.Minute,
+	}
+	qs, err := qsf.NewQueueSet(config)
+	if err != nil {
+		t.Fatalf("QueueSet creation failed with %v", err)
+	}
+
+	exerciseQueueSetUniformScenario(t, qs, []uniformClient{
+		{1001001001, 5, 10, time.Second, time.Second},
+		{2002002002, 5, 10, time.Second, time.Second},
+	}, time.Second*10, true, true, clk, counter)
+}
+
+func TestDifferentFlows(t *testing.T) {
+	now := time.Now()
+
+	clk, counter := clock.NewFakeEventClock(now, 0, nil)
+	qsf := NewQueueSetFactory(clk, counter)
+	config := fq.QueueSetConfig{
+		Name:             "TestDifferentFlows",
+		ConcurrencyLimit: 1,
+		DesiredNumQueues: 128,
+		QueueLengthLimit: 128,
+		HandSize:         1,
+		RequestWaitLimit: 10 * time.Minute,
+	}
+	qs, err := qsf.NewQueueSet(config)
+	if err != nil {
+		t.Fatalf("QueueSet creation failed with %v", err)
+	}
+
+	exerciseQueueSetUniformScenario(t, qs, []uniformClient{
+		{1001001001, 5, 10, time.Second, time.Second},
+		{2002002002, 2, 5, time.Second, time.Second / 2},
+	}, time.Second*10, true, true, clk, counter)
+}
+
+func TestTimeout(t *testing.T) {
+	now := time.Now()
+
+	clk, counter := clock.NewFakeEventClock(now, 0, nil)
+	qsf := NewQueueSetFactory(clk, counter)
+	config := fq.QueueSetConfig{
+		Name:             "TestTimeout",
+		ConcurrencyLimit: 1,
+		DesiredNumQueues: 128,
+		QueueLengthLimit: 128,
+		HandSize:         1,
+		RequestWaitLimit: 0,
+	}
+	qs, err := qsf.NewQueueSet(config)
+	if err != nil {
+		t.Fatalf("QueueSet creation failed with %v", err)
+	}
+
+	exerciseQueueSetUniformScenario(t, qs, []uniformClient{
+		{1001001001, 5, 100, time.Second, time.Second},
+	}, time.Second*10, true, false, clk, counter)
+}
```

Comment:
Do we test changing config anywhere in this file? That seems important, both adding and removing queues, and whatever else is needed.

I would feel a lot more confident about the locking / counting if we could make a "thrash" test, e.g. use a real clock and simulate a bunch of random requests.

Reply:
Yes, a randomized tester with config changes would be a good add.  It would not need to use a real clock.

Reply:
I think some testing with a real clock adds assurance that there's no deadlocks hiding in the code.